// @shopify/polaris/esnext/components/Connected/components/Item/Item.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { autobind } from '@shopify/javascript-utilities/decorators';
import { classNames } from '@shopify/react-utilities/styles';
import styles from '../../Connected.scss';
export var ItemPosition;
(function (ItemPosition) {
    ItemPosition[ItemPosition["Left"] = 0] = "Left";
    ItemPosition[ItemPosition["Primary"] = 1] = "Primary";
    ItemPosition[ItemPosition["Right"] = 2] = "Right";
})(ItemPosition || (ItemPosition = {}));
export default class Item extends React.PureComponent {
    constructor() {
        super(...arguments);
        this.state = { focused: false };
    }
    render() {
        const { focused } = this.state;
        const { children, position } = this.props;
        const className = classNames(styles.Item, focused && styles['Item-focused'], position === ItemPosition.Primary
            ? styles['Item-primary']
            : styles['Item-connection']);
        return (<div onBlur={this.handleBlur} onFocus={this.handleFocus} className={className}>
        {children}
      </div>);
    }
    handleBlur() {
        this.setState({ focused: false });
    }
    handleFocus() {
        this.setState({ focused: true });
    }
}
tslib_1.__decorate([
    autobind
], Item.prototype, "handleBlur", null);
tslib_1.__decorate([
    autobind
], Item.prototype, "handleFocus", null);
