// @shopify/polaris/esnext/components/TopBar/components/UserMenu/context/Provider.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import isEqual from 'lodash/isEqual';
import { autobind } from '@shopify/javascript-utilities/decorators';
import UserMenuContext from './context';
class Provider extends React.Component {
    constructor() {
        super(...arguments);
        this.state = {
            // eslint-disable-next-line react/no-unused-state
            mobileView: this.props.mobileView,
            mobileUserMenuProps: undefined,
            // eslint-disable-next-line react/no-unused-state
            setMobileUserMenuProps: this.setMobileUserMenuProps,
        };
    }
    static getDerivedStateFromProps({ mobileView: nextMobileView }, { mobileView }) {
        if (nextMobileView !== mobileView) {
            return { mobileView: nextMobileView };
        }
        return null;
    }
    render() {
        const { state } = this;
        const { children } = this.props;
        return (<UserMenuContext.Provider value={state}>
        {children}
      </UserMenuContext.Provider>);
    }
    setMobileUserMenuProps(mobileUserMenuProps) {
        const { mobileUserMenuProps: prevMobileUserMenuProps } = this.state;
        if (isEqual(mobileUserMenuProps, prevMobileUserMenuProps)) {
            return;
        }
        this.setState({ mobileUserMenuProps });
    }
}
tslib_1.__decorate([
    autobind
], Provider.prototype, "setMobileUserMenuProps", null);
export default Provider;
